Create a new check to look for mis-use in calls that take iterators

njames93 opened the pull request:
Looks for various patterns of functions that take arguments calling begin/end or similar for common mistakes.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/list.rst:
```diff
@@ -400,15 +401,14 @@ Clang-Tidy Checks
    :doc:`bugprone-narrowing-conversions <bugprone/narrowing-conversions>`, :doc:`cppcoreguidelines-narrowing-conversions <cppcoreguidelines/narrowing-conversions>`,
    :doc:`cert-con36-c <cert/con36-c>`, :doc:`bugprone-spuriously-wake-up-functions <bugprone/spuriously-wake-up-functions>`,
    :doc:`cert-con54-cpp <cert/con54-cpp>`, :doc:`bugprone-spuriously-wake-up-functions <bugprone/spuriously-wake-up-functions>`,
+   :doc:`cert-ctr56-cpp <cert/ctr56-cpp>`, :doc:`bugprone-pointer-arithmetic-on-polymorphic-object <bugprone/pointer-arithmetic-on-polymorphic-object>`,
```
thx, for pointing those issues. I will correct them.

njames93 committed:
Create a new check to look for mis-use in calls that take iterators

Looks for various patterns of functions that take arguments calling
begin/end or similar for common mistakes

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/IncorrectIteratorsCheck.cpp:
```diff
@@ -0,0 +1,757 @@
+/// Checks to see if a all the parameters of a template function with a given
+/// index refer to the same type.
+AST_MATCHER_P(FunctionDecl, areParametersSameTemplateType, SVU, Indexes) {
+  const auto *TemplateDecl = Node.getPrimaryTemplate();
```
Please don't use `auto` if type is not spelled explicitly in same statement or iterator. Same in other places.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+Different ranges
+================
+
+Looks for calls where the range for the `begin` argument is different to the
```
```suggestion
Looks for calls where the range for the ``begin`` argument is different to the
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+================
+
+Looks for calls where the range for the `begin` argument is different to the
+`end` argument
```
```suggestion
``end`` argument.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+Mismatched Begin/end
+====================
+
+Looks for calls where the `begin` parameter is passed an `end` argument or vice
```
```suggestion
Looks for calls where the ``begin`` parameter is passed an ``end`` argument or vice
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+=================
+
+Looks for calls to methods on containers that expect an iterator inside the
+container but are given a different container
```
```suggestion
container but are given a different container.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+====================
+
+Looks for calls where the `begin` parameter is passed an `end` argument or vice
+versa
```
```suggestion
versa.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+================
+
+Looks for calls which accept a single output iterator but are passed the end of
+a container
```
```suggestion
a container.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+Reverse Iteration
+=================
+
+The check understands `rbegin` and `rend` and ensures they are in the correct places.
```
```suggestion
The check understands ``rbegin`` and ``rend`` and ensures they are in the correct places.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+  std::find(a.rend(), a.rbegin(), 0); // Arguments are swapped.
+
+Manually creating a reverse iterator using the ``std::make_reverse_iterator`` is
+also supported, In this case the check looks for calls to `end` for the `begin`
```
```suggestion
also supported, In this case the check looks for calls to ``end`` for the ``begin``
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/incorrect-iterators.rst:
```diff
@@ -0,0 +1,121 @@
+Manually creating a reverse iterator using the ``std::make_reverse_iterator`` is
+also supported, In this case the check looks for calls to `end` for the `begin`
+parameter and vice versa. The name of functions for creating reverse iterator
+can be configured with the option :option:`MakeReverseIterator`
```
```suggestion
can be configured with the option :option:`MakeReverseIterator`.
```

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectIteratorsCheck.cpp:
```diff
@@ -0,0 +1,756 @@
+          ToRefs({MultiRangePolicySingleOutputIterator}), 5, State),
+      this);
+
+  Finder->addMatcher(
+      makeContainerPairRangeMatcher({"::std::vector", "::std::list"},
+                                    {"insert"}, 1, 2, State),
+      this);
+  Finder->addMatcher(
+      makeContainerInternalMatcher({"::std::vector", "::std::list"},
+                                   {"insert", "erase", "emplace"}, 0, State),
+      this);
```
There are some more container member that expect an iterator of the container, such as `std::deque::erase`. Can we add them now?

njames93 replied:
Yeah, this is a little sore point, that there's not yet(#100349) a nice way to specify the available permutations

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/IncorrectIteratorsCheck.cpp:
```diff
@@ -0,0 +1,756 @@
+      this);
+}
+
+void IncorrectIteratorsCheck::check(const MatchFinder::MatchResult &Result) {
```
Maybe use early return instead of if-else chains?

5chmidti reviewed (commented):
Some initial thoughts, and at first glance, this looks nice. Because we are early in the release cycle, there will be enough time to address the fixme's.

(I haven't run the check against some code bases as I am not at my main machine for another week.)

njames93 commented:
Added some detection for trying to advance past the end or before the start of a range.

njames93 commented:
@5chmidti I decided to run this version of the check on llvm and clang
https://gist.github.com/njames93/6f8863573972cb07c388ac19e8a20467

Seems like most of the warnings detected are when we are deliberately exceeding the bounds of a range when we know its a slice of a bigger range. Some of them are potentially cases where code could be refactored to better express it's intent, other instances already have comments to explain why we are exceeding the bounds

There were also some cases where classes had begin/end pairs that didn't return iterators, instead [just integral types](https://github.com/llvm/llvm-project/blob/main/llvm/lib/DWARFLinker/Parallel/DWARFLinkerCompileUnit.cpp#L682).
These cases could be silenced by changing the check to ensure we aren't looking for integral types, but I feel that's such an unlikely case that there's no need to worry about it

5chmidti commented:
The `++` and `--` operators could be mis-used as well, similar to `+` and `-`. And there are some control statements with a single statement that has a compound statement around it, please remove those.

Checking the log, it looks like the check works well.

I know that TBB has `begin` and `end` members that return integral types. But I'm not sure that we need to exclude checking for integral types. E.g., `a.end() + 1` could still be bupgrone, even if `end` returns an `int`.

njames93 commented:
> The `++` and `--` operators could be mis-used as well, similar to `+` and `-`. And there are some control statements with a single statement that has a compound statement around it, please remove those.

The `++` and `--` operators work on lvalues only(`std::end(Range)++` is usually a compiler error.), Therefore, in order to detect that, there would need to be some control flow logic in the check, Something along the lines of detecting setting a variable to `Range.end()` then looking for any calls to `operator++` on that variable. This logic definitely could be done and the results could catch many more questionable cases, but for the first version of this check I think it can be left out.

5chmidti reviewed (approved):
LGTM (- single statement compound statements), but wait for another review please

njames93 commented:
- Cleaned up handling of functions that take an execution policy as first argument.
- Handle `std::trasnform` and its overloads as a special case(Its overloads are a bit of a pain)

whisperity commented on clang-tools-extra/clang-tidy/bugprone/IncorrectIteratorsCheck.h:
```diff
@@ -0,0 +1,45 @@
+  bool isLanguageVersionSupported(const LangOptions &LangOpts) const override;
+
+private:
+  std::vector<StringRef> BeginFree;
+  std::vector<StringRef> EndFree;
+  std::vector<StringRef> BeginMethod;
+  std::vector<StringRef> EndMethod;
+  std::vector<StringRef> RBeginFree;
+  std::vector<StringRef> REndFree;
+  std::vector<StringRef> RBeginMethod;
+  std::vector<StringRef> REndMethod;
+  std::vector<StringRef> MakeReverseIterator;
```
These are usually `const`, as these correspond to the options obtained from the user's configuration.

EugeneZelenko commented:
Please rebase from `main`.
